sno/working_copy/base.py

```diff
@@ -39,18 +44,104 @@ def __str__(self):
 
 
 class WorkingCopy:
+    """
+    Abstract working copy implementation.
+    Subclasses to override any unimplemented methods below, and also to set the following fields:
+
+    self.repo - SnoRepo containing this WorkingCopy
+    self.path - string describing the location of this WorkingCopy
+    self.engine - sqlalchemy engine for connecting to the database
+    self.sessionmaker - sqlalchemy sessionmaker bound to the engine
+    self.preparer - sqlalchemy IdentifierPreparer for quoting SQL in the appropriate dialect
+
+    self.db_schema - database-schema that this working copy controls, if any.
+    self.sno_tables - sqlalchemy Table definitions for sno_track and sno_state tables.
+    """
+
     SNO_WORKINGCOPY_PATH = "sno.workingcopy.path"
 
-    TRACKING_NAME = "track"
-    STATE_NAME = "state"
+    @property
+    @functools.lru_cache(maxsize=1)
+    def DB_SCHEMA(self):
+        """Escaped, dialect-specific name of the database-schema owned by this working copy (if any)."""
+        if self.db_schema is None:
+            raise RuntimeError("No schema to escape.")
+        return self.preparer.format_schema(self.db_schema)
 
     @property
-    def TRACKING_TABLE(self):
-        return self._sno_table(self.TRACKING_NAME)
+    @functools.lru_cache(maxsize=1)
+    def SNO_TRACK(self):
+        """Escaped, dialect-specific fully-qualified name of sno_track table."""
+        return self.table_identifier(self.sno_tables.sno_track)
 
     @property
-    def STATE_TABLE(self):
-        return self._sno_table(self.STATE_NAME)
+    @functools.lru_cache(maxsize=1)
+    def SNO_STATE(self):
+        """Escaped, dialect-specific fully-qualified name of sno_state table."""
+        return self.table_identifier(self.sno_tables.sno_state)
+
+    @property
+    @functools.lru_cache(maxsize=1)
+    def SNO_TRACK_NAME(self):
+        """The table name of sno_track table, not including the schema."""
+        return self.sno_tables.sno_track.name
+
+    @property
+    @functools.lru_cache(maxsize=1)
+    def SNO_STATE_NAME(self):
+        """The table name of sno_track table, not including the schema."""
+        return self.sno_tables.sno_state.name
+
+    @property
+    def CAST_TO_TEXT(self):
+        """Returns a SQL suffix that casts an expression to text."""
```

Comment:
postgres' `::TEXT` is a bit non-standard, the standard SQL way is `CAST(fieldname AS TEXT)`, which afaik works in both sqlite and postgres. But sqlalchemy also has a way of producing it: https://docs.sqlalchemy.org/en/13/core/sqlelement.html#sqlalchemy.sql.expression.cast

So... the existence of this property seems odd/unnecessary

Reply:
Removed, added `CAST(fieldname AS TEXT)`